bugfix - check config file existence for basicAuth filter

ayruslore opened the pull request:
filters/auth: check config file existence for basicAuth filter

Missing config file causes basicAuth filter to panic.
This change adds file existence check during filter creation.

Fixes https://github.com/zalando/skipper/issues/3327

AlexanderYastrebov commented on filters/auth/basic.go:
```diff
@@ -65,6 +75,13 @@ func (spec *basicSpec) CreateFilter(config []interface{}) (filters.Filter, error
 		}
 	}
 
+	if _, err := os.Stat(configFile); err != nil {
+		return &basic{
+			authenticator:   nil,
+			realmDefinition: ForceBasicAuthHeaderValue + `"` + realmName + `"`,
+		}, nil
+	}
```
I think we should fail early here if file does not exist (and the datasource will discard the whole route).
If file was removed later in runtime then its fine to return 500 like it is now.

ayruslore replied:
> I think we should fail early here if file does not exist (and the datasource will discard the whole route)

But If the route is discarded then the response would be 404 instead of 401. What do we want as expected behaviour?

To return a 401 we need to match the route and then fail the auth, and if we discard the route, then this will never match returning 404 or status code something other route that matches....

> If file was removed later in runtime then its fine to return 500 like it is now.

this makes sense.

AlexanderYastrebov replied:
We need to fail somehow as current approach simply masks the problem and user/operator won't tell why they get 401.

ayruslore replied:
i added a body message to show to the user. See line https://github.com/zalando/skipper/pull/3467/files#diff-eef77465b5f737710ce3478e21e86d18fa5ce13979926f4f19cd38a8457d6e5dR41

Now with a 401 we indicate that the credential file is missing. Would that suffice or do you have any other suggestions indicating why the 401?

This is what current response looks like now.
```
❯ curl -v test:test@localhost:9090
* Host localhost:9090 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:9090...
* Connected to localhost (::1) port 9090
* Server auth using Basic with user 'test'
> GET / HTTP/1.1
> Host: localhost:9090
> Authorization: Basic dGVzdDp0ZXN0
> User-Agent: curl/8.7.1
> Accept: */*
> 
* Request completely sent off
< HTTP/1.1 401 Unauthorized
< Server: Skipper
< Date: Fri, 04 Apr 2025 12:38:51 GMT
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
missing credential file%    
```

ayruslore replied:
Update : I have simplified and now it fails early in Create Filter and ignores the route. Consistent with the behaviour of other filters.

MustafaSaber commented:
I think last commit wasn't supposed to be on this branch

ayruslore commented:
> I think last commit wasn't supposed to be on this branch

Yes. I have removed that commit. I pushed to the wrong branch. Now the commit is excluded from here and added to the right place.

ayruslore commented on filters/auth/basic.go:
```diff
@@ -65,6 +68,10 @@ func (spec *basicSpec) CreateFilter(config []interface{}) (filters.Filter, error
 	}
 
+	if _, err := os.Stat(configFile); err != nil {
+		return nil, errors.New("htpasswd file not found: " + configFile)
```
done. updated.

ayruslore commented on filters/auth/basic_test.go:
```diff
@@ -71,6 +71,14 @@ func TestWithSuccessfulAuth(t *testing.T) {
+func TestWithMissingAuthFile(t *testing.T) {
+	spec := NewBasicAuth()
+	_, err := spec.CreateFilter([]interface{}{"testdata/missingfile"})
+	if err == nil {
```
fixed.

AlexanderYastrebov commented on filters/auth/basic.go:
```diff
@@ -65,6 +68,10 @@ func (spec *basicSpec) CreateFilter(config []interface{}) (filters.Filter, error
 	}
 
+	if _, err := os.Stat(configFile); err != nil {
+		return nil, fmt.Errorf("htpasswd file not found: %s", configFile)
```
Lets wrap error here and also make message generic since we don't know what the problem exactly is:
```go
return nil, fmt.Errorf("stat failed for %q: %w", configFile, err)
```

ayruslore replied:
updated the code.

MustafaSaber commented on filters/auth/basic_test.go:
```diff
@@ -71,6 +72,13 @@ func TestWithSuccessfulAuth(t *testing.T) {
 	}
 }
 
+func TestWithMissingAuthFile(t *testing.T) {
```
I think this can be added under `TestCreateFilterBasicAuthErrorCases`, but can also be done in another PR.
